### PR TITLE
TaskManager Community Module

### PR DESCRIPTION
**Pull request by NielsCharlier**
This new community module was approved on the mailing list: http://osgeo-org.1560.x6.nabble.com/new-community-module-workflow-td5321612.html

Extensive documentation and further improvements are on its way.

The main purpose of this PR is for someone to approve the very minimal changes in core.

### CODE REVIEW

File: src/community/pom.xml [hunk 1]
```diff
@@ -230,7 +230,7 @@
         <module>gwc-distributed</module>
         <module>jdbcstore</module>
         <module>gdal</module>
-        <module>gwc-s3</module>
+        <!-- <module>gwc-s3</module> -->
```
Comment (bradh) [inline src/community/pom.xml]: Why this change?

File: src/community/taskmanager/pom.xml [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+                                <artifactId>postgresql</artifactId>
+                                <version>8.1-407.jdbc3</version>
+                            </dependency>
+                        </dependencies>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+        <profile>
+            <id>myotherdb</id>
```
Comment (bradh) [inline src/community/taskmanager/pom.xml]: Probably true, but maybe you can find a more descriptive title?

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Attribute.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    
+    String getValue();
+    
+    void setValue(String value);
+    
+    Configuration getConfiguration();
+    
+    void setConfiguration(Configuration configuration);
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Attribute.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Batch.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    
+    List<BatchRun> getBatchRuns();
+    
+    default String getFullName() {
+        return getConfiguration() == null ? getName() : 
+            ((getConfiguration().getName() == null ? "" : 
+                getConfiguration().getName()) + FULL_NAME_DIVISOR + getName());
+    }
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Batch.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/BatchElement.java [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    void setTask(Task task);
+
+    Integer getIndex();
+
+    void setIndex(Integer index);
+
+    List<Run> getRuns();
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/BatchElement.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Configuration.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+    String getName();
+
+    void setName(String name);
+
+    String getDescription();
+
+    void setDescription(String name);
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Configuration.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Parameter.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+    String getValue();
+
+    void setValue(String value);
+
+    Task getTask();
+
+    void setTask(Task task);
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Parameter.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Run.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    
+    String getMessage();
+
+    void setMessage(String message);
+    
+    BatchRun getBatchRun();
+    
+    void setBatchRun(BatchRun br);
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Run.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Task.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    String getName();
+
+    void setName(String name);
+    
+    default String getFullName() {
+        return (getConfiguration().getName() == null ? "" : getConfiguration().getName()) +                 
+                FULL_NAME_DIVISOR + getName();
+    }
+
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/data/Task.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/external/DefaultDialectImpl.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.Set;
+
+import org.geoserver.taskmanager.util.SqlUtil;
+
+/**
+ * Default implementation for the Dialect interface.
+ * 
+ * This should work with most databases, but it also limits the functionality of the task manager.
+ * 
+ * @author Timothy De Bock
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/external/DefaultDialectImpl.java]: Are you contributing on behalf of this author?

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/Report.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+    public String getContent() {
+        return content;
+    }
+    
+    public Type getType() {
+        return type;
+    }
+    
+}
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/Report.java]: Needs EOL.

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/impl/SimpleReportBuilderImpl.java [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+/**
+ * A very simple report builder.
+ * Contains all information that matters in simple text format.
+ * 
+ * @author Niels Charlier
+ *
+ */
+@Service
+public class SimpleReportBuilderImpl implements ReportBuilder {
+
```
Comment (bradh) [inline src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/impl/SimpleReportBuilderImpl.java]: Consider adding logging for failure cases here?

File: src/community/pom.xml (hunk 1 above)
Comment (NielsCharlier) [reply src/community/pom.xml]: hmmm that is a mistake, needs to be undone

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/impl/SimpleReportBuilderImpl.java (hunk 12 above)
Comment (NielsCharlier) [reply src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/impl/SimpleReportBuilderImpl.java]: not sure what you mean... exceptions and batch start/stop/cancel events are sent to the general log in BatchJobImpl . If you want to send information about batch results elsewhere you can write your own reportservice

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/external/DefaultDialectImpl.java (hunk 10 above)
Comment (NielsCharlier) [reply src/community/taskmanager/src/main/java/org/geoserver/taskmanager/external/DefaultDialectImpl.java]: yes

File: src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/impl/SimpleReportBuilderImpl.java (hunk 12 above)
Comment (bradh) [reply src/community/taskmanager/src/main/java/org/geoserver/taskmanager/report/impl/SimpleReportBuilderImpl.java]: OK. I thought it might be useful to provide some additional logging context information here, but it was a "maybe" level suggestion.

**Comment by NielsCharlier** [comment]
Here are direct links to the two small changes in the main module:

https://github.com/geoserver/geoserver/pull/2759/commits/6945957a93b326ddaa39ee00c3ca4a707d663aee#diff-ed7269d9c2be90db1cbfde7d0f32d751

https://github.com/geoserver/geoserver/pull/2759/commits/6945957a93b326ddaa39ee00c3ca4a707d663aee#diff-5fac1624412cebb8475c6e0fee6fd17c

File: src/main/src/test/java/org/geoserver/data/test/MockData.java [hunk 13]
```diff
@@ -56,6 +57,8 @@
  * @author Justin Deoliveira, The Open Planning Project
  *
  */
+
```
Comment (bencaradocdavies) [inline src/main/src/test/java/org/geoserver/data/test/MockData.java]: @NielsCharlier there should be no blank lines between the class Javadoc and the class definition. Other than this, your changes to gs-main and the gs-web-app pom look fine to me.

**Comment by NielsCharlier** [comment]
Thanks @bencaradocdavies

**Comment by aaime** [comment]
The merge broke the nightly build:

````
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.3.2:testCompile (default-testCompile) on project gs-taskmanager: Compilation failure
[ERROR] /var/jenkins/workspace/geoserver-master/src/community/taskmanager/src/test/java/org/geoserver/taskmanager/tasks/DbRemotePublicationTaskTest.java:[85,22] error: cannot find symbol
[ERROR] -> [Help 1]
````
More info here:
https://build.geoserver.org/view/geoserver/job/geoserver-master-nightly/1727/console

I've commented out the module from the "communityRelease" profile, please fix and add back as you see fit.

**Comment by aaime** [comment]
Actually, it broke also regular builds:

image
